wallet-new/src/Cardano/Wallet/Kernel/Decrypt.hs

```diff
@@ -2,27 +2,69 @@ module Cardano.Wallet.Kernel.Decrypt
     ( decryptAddress
     , decryptHdLvl2DerivationPath
     , keyToWalletDecrCredentials
+    , selectOwnAddresses
     , WalletDecrCredentials
-    , WalletDecrCredentialsKey(..)
+    , WalletDecrCredentialsKey (..)
     ) where
 
 import           Universum
 
-import           Data.List ((!!))
-
-import           Pos.Wallet.Web.Tracking.Decrypt
+import           Formatting (build, sformat)
 
+import           Cardano.Wallet.API.V1.Types as V1
 import           Cardano.Wallet.Kernel.DB.HdWallet (HdAccountIx (..),
                      HdAddressIx (..))
-import           Pos.Core (Address, aaPkDerivationPath, addrAttributesUnwrapped)
-import           Pos.Crypto (HDPassphrase, unpackHDAddressAttr)
-
+import           Pos.Core (aaPkDerivationPath, addrAttributesUnwrapped,
+                     makeRootPubKeyAddress)
+import           Pos.Crypto (EncryptedSecretKey, HDPassphrase, PublicKey,
+                     deriveHDPassphrase, encToPublic, unpackHDAddressAttr)
 
 decryptHdLvl2DerivationPath :: HDPassphrase
                             -> Address
                             -> Maybe (HdAccountIx, HdAddressIx)
 decryptHdLvl2DerivationPath hdPass addr = do
     hdPayload <- aaPkDerivationPath $ addrAttributesUnwrapped addr
     derPath <- unpackHDAddressAttr hdPass hdPayload
-    guard $ length derPath == 2
-    pure (HdAccountIx (derPath !! 0), HdAddressIx (derPath !! 1))
+    case derPath of
+        [a,b] -> Just (HdAccountIx a, HdAddressIx b)
+        _     -> Nothing
+
+
+
+type WalletDecrCredentials = (HDPassphrase, V1.WalletId)
+
+-- | Key to identify regular or external wallet.
+data WalletDecrCredentialsKey
+    = KeyForRegular EncryptedSecretKey
+    | KeyForExternal PublicKey
+    deriving (Show)
+
+-- | There's a secret key for regular wallet or a public key for external wallet.
+keyToWalletDecrCredentials :: WalletDecrCredentialsKey -> WalletDecrCredentials
+keyToWalletDecrCredentials (KeyForRegular sk)  = credentialsFromPublicKey $ encToPublic sk
+keyToWalletDecrCredentials (KeyForExternal pk) = credentialsFromPublicKey pk
+
+credentialsFromPublicKey :: PublicKey -> WalletDecrCredentials
+credentialsFromPublicKey publicKey = (hdPassword, walletId)
+  where
+    hdPassword = deriveHDPassphrase publicKey
+
+    -- When migrating from 'Pos.Wallet.Web.Tracking.Decrypt' this type is
+    -- changed from 'CId Wal' to 'V1.WalletId'
+    walletId   = V1.WalletId . (sformat build) $ makeRootPubKeyAddress publicKey
```

Comment:
What would be nice here is to add a small unit test to ensure we aren't breaking the representation. 

We use to go back and forth between types via:
```hs
instance Migrate (V0.CId V0.Wal) V1.WalletId where
    eitherMigrate (V0.CId (V0.CHash h)) = pure (V1.WalletId h)
```

So, it's unlikely we broke something. Still, it would feel safe to convince ourselves via a simple test.